### PR TITLE
Tech/move filestates to redux

### PR DESCRIPTION
**Pull request by NearW**
# Move the filestates to the redux store

## Description

This PR should prevent the fileStates from being mutated unpredictably by moving it to the redux store and only mutating it with redux actions. This should make the dataflow easier and also make future performance issues (cloning a lot) easier to solve.

- Created a new folder `model`
- Added class `Files` that handles all reducer functions and provides the functionality of the previous fileStateHelper class
- Refactored some functions that passed a parameter through and through instead of just reading it from the store, where the parameter was needed

### Notes to reviewer
- When everything is reviewed and approved, we should move the codeCharta.model.ts into the model folder
- Since this was a major refactoring, please take your time to give it a long manual test

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

- [ ] **All** requirements mentioned in the issue are implemented
- [x] Does match the Code of Conduct and the Contribution file
- [ ] Task has its own **GitHub issue** (something it is solving)
  - [ ] Issue number is **included in the commit messages** for traceability
- [x] **Update the README.md** with any changes/additions made
- [x] **Update the CHANGELOG.md** with any changes/additions made
- [x] **Enough test coverage to ensure that uncovered changes do not break functionality**
- [x] **All tests pass**
- [x] **Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?
- [x] **Assign your PR to someone** for a code review
  - This person _will be contacted **first**_ if a bug is introduced into `master`
- [x] **Manual testing** did not fail

### CODE REVIEW

**Comment by alschmut** [comment]
@NearW are you sure to add the files (old: fileStates) to the fileSettings? Because to in fact, a CCFile can contain fileSettings, which makes it recursive then. I would rather put them side by side with the fileSettings/dynamicSettings etc. on the top level.

**Comment by NearW** [comment]
> @NearW are you sure to add the files (old: fileStates) to the fileSettings? Because in fact, a CCFile can contain fileSettings, which makes it recursive then. I would rather put them side by side with the fileSettings/dynamicSettings etc. on the top level.

I know. See the first todo checkbox. :D

File: visualization/app/codeCharta/state/store/fileSettings/markedPackages/markedPackages.service.spec.ts [hunk 1]
```diff
@@ -66,4 +66,6 @@ describe("MarkedPackagesService", () => {
 			expect($rootScope.$broadcast).not.toHaveBeenCalled()
 		})
 	})
+
+	describe("onFilesChanged", () => {})
```
Comment (JaiJoPa) [inline visualization/app/codeCharta/state/store/fileSettings/markedPackages/markedPackages.service.spec.ts]: Do we need this ?

**Comment by BridgeAR** [comment]
This needs a rebase with master (`git fetch && git rebase -i origin/master --autostash && git push --force-with-lease`).

The PR itself is pretty huge and very difficult to review due to that. Is it possible to split the PR into multiple smaller chunks that can be reviewed independently from each other?

File: visualization/app/codeCharta/state/store/fileSettings/markedPackages/markedPackages.service.spec.ts (hunk 1 above)
Comment (NearW) [reply visualization/app/codeCharta/state/store/fileSettings/markedPackages/markedPackages.service.spec.ts]: No, I wanted to add a test for this method, but the merging methods are tested in their own file already. Perhaps we can remove the merger tests and add them to this method here.

Removed this block for now though. :)

**Comment by NearW** [comment]
@BridgeAR I reverted the relocation of the model file again. The PR is still pretty big, but it's mostly replacing methods and adapting tests and doing some boyscouting here and there. 

If this PR is too big to review, I can do a review with one of you together.

File: visualization/app/codeCharta/state/metric.service.ts [hunk 2]
```diff
@@ -157,12 +160,13 @@ export class MetricService implements FileStateSubscriber, BlacklistSubscriber {
...
+	private getUniqueMetricNames(): string[] {
+		const visibleFileStates = this.storeService.getState().files.getVisibleFileStates()
+		if (visibleFileStates.length === 0) {
 			return []
 		} else {
 			let leaves: HierarchyNode<CodeMapNode>[] = []
-			fileStates.forEach((fileState: FileState) => {
+			visibleFileStates.forEach((fileState: FileState) => {
 				leaves = leaves.concat(hierarchy<CodeMapNode>(fileState.file.map).leaves())
 			})
```
Comment (alschmut) [inline visualization/app/codeCharta/state/metric.service.ts]: How about following:
```
const files = this.storeService.getState().files
if (files.fileStatesAvailable()) {
    ... // files.getVisibleFiles()...
} else {
    return []
}
```

File: visualization/app/codeCharta/state/store/dynamicSettings/focusedNodePath/focusedNodePath.service.spec.ts [hunk 3]
```diff
@@ -79,11 +79,11 @@ describe("FocusedNodePathService", () => {
 		})
 	})
 
-	describe("onFileStatesChanged", () => {
+	describe("onFilesChanged", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/state/store/dynamicSettings/focusedNodePath/focusedNodePath.service.spec.ts]:
```suggestion
	describe("onFilesSelectionChanged", () => {
```

File: visualization/app/codeCharta/state/store/dynamicSettings/searchPattern/searchPattern.service.spec.ts [hunk 4]
```diff
@@ -67,11 +68,11 @@ describe("SearchPatternService", () => {
 		})
 	})
 
-	describe("onFileStatesChanged", () => {
+	describe("onFilesChanged", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/state/store/dynamicSettings/searchPattern/searchPattern.service.spec.ts]:
```suggestion
	describe("onFilesSelectionChanged", () => {
```

File: visualization/app/codeCharta/state/store/fileSettings/markedPackages/markedPackages.service.ts [hunk 5]
```diff
@@ -25,13 +25,13 @@ export class MarkedPackagesService implements StoreSubscriber, FileStateSubscrib
...
-	public onFileStatesChanged(fileStates: FileState[]) {
+	public onFilesSelectionChanged(fileStates: Files) {
 		this.merge(fileStates)
 	}
 
-	private merge(fileStates: FileState[]) {
-		const visibleFiles = FileStateHelper.getVisibleFileStates(fileStates).map(x => x.file)
-		const withUpdatedPath = FileStateHelper.isPartialState(fileStates)
+	private merge(files: Files) {
+		const visibleFiles = files.getVisibleFileStates().map(x => x.file)
```
Comment (alschmut) [inline visualization/app/codeCharta/state/store/fileSettings/markedPackages/markedPackages.service.ts]:
```suggestion
		const visibleFiles = files.getVisibleFiles()
```

File: visualization/app/codeCharta/state/store/files/files.reducer.spec.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+			const result = files(state, setFiles(newFiles))
+
+			expect(result).toEqual(newFiles)
+		})
+
+		it("should set default files", () => {
+			const result = files(state, setFiles())
+
+			expect(result).toEqual(new Files())
```
Comment (alschmut) [inline visualization/app/codeCharta/state/store/files/files.reducer.spec.ts]: Do you wanna use the imported `defaultFiles` here like above?
```suggestion
			expect(result).toEqual(defaultFiles)
```

File: visualization/app/codeCharta/state/store/files/files.reducer.spec.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+			expect(result.getFiles().length).toBe(0)
+		})
+	})
+
+	describe("Action: RESET_SELECTION", () => {
+		it("should clear and reset the state", () => {
+			const result = files(state, resetFiles())
+
+			expect(result.getFiles().length).toBe(0)
```
Comment (alschmut) [inline visualization/app/codeCharta/state/store/files/files.reducer.spec.ts]: Copy paste mistake: Use `resetSelection()` here and check with `fileStatesAvailable()`.

File: visualization/app/codeCharta/state/store/files/files.reducer.spec.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	describe("Action: SET_SINGLE_BY_NAME", () => {
+		it("should select a file by name to view in single mode", () => {
+			const result = files(state, setSingleByName(TEST_DELTA_MAP_A.fileMeta.fileName))
+
+			expect(result.isSingleState()).toBeTruthy()
+		})
+	})
+
+	describe("Action: SET_DELTA", () => {
+		it("should select a file a file as reference and another as comparison", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/state/store/files/files.reducer.spec.ts]:
```suggestion
		it("should select a file as reference and another as comparison", () => {
```

File: visualization/app/codeCharta/state/store/files/files.reducer.spec.ts [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+			const result = files(state, setDelta(TEST_DELTA_MAP_A, TEST_DELTA_MAP_B))
+
+			expect(result.isDeltaState()).toBeTruthy()
+		})
+	})
+
+	describe("Action: SET_DELTA_BY_NAMES", () => {
+		it("should select a file a file as reference and another as comparison by name", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/state/store/files/files.reducer.spec.ts]:
```suggestion
		it("should select a file as reference and another as comparison by name", () => {
```

File: visualization/app/codeCharta/state/store/files/files.reducer.spec.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+			const result = files(state, setDeltaByNames(TEST_DELTA_MAP_A.fileMeta.fileName, TEST_DELTA_MAP_B.fileMeta.fileName))
+
+			expect(result.isDeltaState()).toBeTruthy()
+		})
+	})
+
+	describe("Action: SET_MULTIPLE", () => {
+		it("should select a file a file as reference and another as comparison", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/state/store/files/files.reducer.spec.ts]:
```suggestion
		it("should select two files to view in multiple mode", () => {
```

File: visualization/app/codeCharta/state/store/files/files.splitter.ts [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import { setFiles } from "./files.actions"
+import { CCAction } from "../../../codeCharta.model"
+import { Files } from "../../../model/files"
+
+export function splitFilesAction(payload: Files): CCAction[] {
+	return [setFiles(payload)]
+}
```
Comment (alschmut) [inline visualization/app/codeCharta/state/store/files/files.splitter.ts]: Why do we need to return an array of CCActions..? Might be correct, but I need some clarification :)

File: visualization/app/codeCharta/ui/areaSettingsPanel/areaSettingsPanel.component.spec.ts [hunk 12]
```diff
@@ -75,15 +75,15 @@ describe("AreaSettingsPanelController", () => {
 		})
 	})
 
-	describe("onFileStatesChanged", () => {
+	describe("onFilesChanged", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/ui/areaSettingsPanel/areaSettingsPanel.component.spec.ts]:
```suggestion
	describe("onFilesSelectionChanged", () => {
```

File: visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.spec.ts [hunk 13]
```diff
@@ -196,6 +205,14 @@ describe("codeMapMouseEventService", () => {
 		})
 	})
 
+	describe("onFilesChanged", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.spec.ts]:
```suggestion
	describe("onFilesSelectionChanged", () => {
```

File: visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts [hunk 14]
```diff
@@ -83,21 +80,26 @@ export class CodeMapPreRenderService implements StoreSubscriber, MetricServiceSu
 	}
 
 	private updateRenderMapAndFileMeta() {
-		const unifiedFile: CCFile = this.getSelectedFilesAsUnifiedMap()
+		const unifiedFile = this.getSelectedFilesAsUnifiedMap()
```
Comment (alschmut) [inline visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts]: Why did you remove the type?
```suggestion
		const unifiedFile: CCFile = this.getSelectedFilesAsUnifiedMap()
```

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.spec.ts [hunk 15]
```diff
@@ -93,15 +94,23 @@ describe("ThreeOrbitControlsService", () => {
...
+			FilesService.subscribe = jest.fn()
+
+			rebuildService()
+
+			expect(FilesService.subscribe).toHaveBeenCalledWith($rootScope, threeOrbitControlsService)
+		})
 	})
 
-	describe("onFileStateChanged", () => {
+	describe("onFilesChanged", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.spec.ts]:
```suggestion
	describe("onFilesSelectionChanged", () => {
```

File: visualization/app/codeCharta/ui/colorSettingsPanel/colorSettingsPanel.component.spec.ts [hunk 16]
```diff
@@ -108,25 +115,19 @@ describe("ColorSettingsPanelController", () => {
 		})
 	})
 
-	describe("onFileStatesChanged", () => {
+	describe("onFilesChanged", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/ui/colorSettingsPanel/colorSettingsPanel.component.spec.ts]:
```suggestion
	describe("onFilesSelectionChanged", () => {
```

File: visualization/app/codeCharta/ui/filePanel/filePanel.component.spec.ts [hunk 17]
```diff
@@ -1,111 +1,82 @@
...
 
-	describe("onFileStatesChanged", () => {
-		beforeEach(() => {
-			FileStateHelper.isSingleState = jest.fn().mockReturnValue(true)
-			FileStateHelper.isPartialState = jest.fn().mockReturnValue(true)
-			FileStateHelper.isDeltaState = jest.fn().mockReturnValue(true)
-			FileStateHelper.getVisibleFileStates = jest.fn().mockReturnValue(fileStates)
-		})
-
+	describe("onFilesChanged", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/ui/filePanel/filePanel.component.spec.ts]:
```suggestion
	describe("onFilesSelectionChanged", () => {
```

File: visualization/app/codeCharta/ui/filePanel/filePanel.component.spec.ts [hunk 18]
```diff
@@ -1,111 +1,82 @@
...
+	function initFiles() {
+		storeService.dispatch(resetFiles())
+		storeService.dispatch(addFile(TEST_DELTA_MAP_A))
+		storeService.dispatch(addFile(TEST_DELTA_MAP_B))
+		storeService.dispatch(setSingle(TEST_DELTA_MAP_A))
+	}
 
-	it("should subscribe to FileStateService on construction", () => {
-		FileStateService.subscribe = jest.fn()
+	it("should subscribe to FilesService on construction", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/ui/filePanel/filePanel.component.spec.ts]: Wrap in `describe("constructor", () => {})`

File: visualization/app/codeCharta/ui/filePanel/filePanel.component.ts [hunk 19]
```diff
@@ -143,11 +144,11 @@ export class FilePanelController implements FileStateSubscriber {
...
 		} else if (this.lastRenderState === FileSelectionState.Partial) {
-			const visibleFileStates = FileStateHelper.getVisibleFileStates(this._viewModel.fileStates)
-			if (FileStateHelper.getVisibleFileStates(this._viewModel.fileStates).length > 0) {
+			const visibleFileStates = this._viewModel.files.getVisibleFileStates()
+			if (visibleFileStates.length > 0) {
 				return visibleFileStates[0].file.fileMeta.fileName
 			} else {
-				return this._viewModel.fileStates[0].file.fileMeta.fileName
+				return this._viewModel.files.getFiles()[0].file.fileMeta.fileName
 			}
```
Comment (alschmut) [inline visualization/app/codeCharta/ui/filePanel/filePanel.component.ts]: Should we use `files.fileStatesAvailable()` here?

File: visualization/app/codeCharta/ui/rangeSlider/rangeSlider.component.spec.ts [hunk 20]
```diff
@@ -267,15 +263,15 @@ describe("RangeSliderController", () => {
 		it("should set maxMetricValue", () => {
```
Comment (alschmut) [inline visualization/app/codeCharta/ui/rangeSlider/rangeSlider.component.spec.ts]: Use `onFilesSelectionChanged` string above in describe name.

File: visualization/app/codeCharta/model/files.ts [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+			}
+		})
+	}
+
+	public setSingle(file: CCFile) {
+		this.setSingleByName(file.fileMeta.fileName)
+	}
+
+	public setDeltaByNames(referenceFileName: string, comparisonFileName: string) {
+		this.files = this.files.map(elem => {
```
[the remark this answers is not in the document]
Comment (NearW) [reply visualization/app/codeCharta/model/files.ts]: Agreed and done.

File: visualization/app/codeCharta/model/files.ts [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+		})
+	}
+
+	public setDelta(reference: CCFile, comparison: CCFile) {
+		this.setDeltaByNames(reference.fileMeta.fileName, comparison.fileMeta.fileName)
+	}
+
+	public setMultipleByNames(partialFileNames: string[]) {
+		this.files = this.files.map(elem => {
+			if (partialFileNames.indexOf(elem.file.fileMeta.fileName) !== -1) {
```
[the remark this answers is not in the document]
Comment (NearW) [reply visualization/app/codeCharta/model/files.ts]: Way better. done.

File: visualization/app/codeCharta/state/edgeMetricData.service.ts [hunk 23]
```diff
@@ -89,38 +80,44 @@ export class EdgeMetricDataService implements FileStateSubscriber, BlacklistSubs
...
 			.map(fileState => CodeMapHelper.getAllPaths(fileState.file.map))
 		const allFilePaths: string[] = [].concat(...pathsPerFileState)
-		fileStates.forEach(fileState => {
-			fileState.file.settings.fileSettings.edges.forEach(edge => {
-				if (this.bothNodesAssociatedAreVisible(edge, blacklist, allFilePaths)) {
-					this.addEdgeToCalculationMap(edge)
-				}
+		this.storeService
+			.getState()
+			.files.getFiles()
```
[the remark this answers is not in the document]
Comment (NearW) [reply visualization/app/codeCharta/state/edgeMetricData.service.ts]: That's a good question. I just reused the existing code and tried to not do refactorings I'm not 100% sure about. I assume that visibleFileStates is correct here.

[the remark this answers is not in the document]
Comment (alschmut) [reply visualization/app/codeCharta/state/edgeMetricData.service.ts]: Well, the function is called with `this.calculateMetrics(visibleFileStates)` and then the parameter is renamed to `private calculateMetrics(visibleFileStates: FileState[])`

File: visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts (hunk 14 above)
Comment (NearW) [reply visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts]: Because the type is given by the return value of the function already. Your IDE should recognise that. No need to specify the type here again.

File: visualization/app/codeCharta/state/store/files/files.splitter.ts (hunk 11 above)
Comment (NearW) [reply visualization/app/codeCharta/state/store/files/files.splitter.ts]: Splitter is required when we want to update the whole state object and split each property to delegate it to their reducer. We implemented it the way, that splitting an action returns an array of CCActions, even though there is only one action when there are no more nested reducers (see blacklist or files).

Comment (alschmut) [reply visualization/app/codeCharta/state/store/files/files.splitter.ts]: oh yea right, because its on the same level as the `dynamicSettings.splitter` etc. Thanks! (:

File: visualization/app/codeCharta/model/files.ts [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+		return matchingFileState ? matchingFileState.file : undefined
+	}
+
+	public getVisibleFiles(): CCFile[] {
+		return this.getVisibleFileStates().map(x => x.file)
+	}
+
+	public getVisibleFileStates(): FileState[] {
+		return this.files.filter(x => x.selectedAs !== FileSelectionState.None)
+	}
```
Comment (alschmut) [inline visualization/app/codeCharta/model/files.ts]: Something else I realised: We have functions called `getFiles()` and `getCCFiles()`, but here the functions are called `getVisibleFiles()` and `getVisibleFileStates()`. Maybe we could try using the same words for the same meaning. Now we have fileStates, files and ccFiles..

Comment (NearW) [reply visualization/app/codeCharta/model/files.ts]: Files should be CCFiles, so that we only have FileStates and CCFiles

**Comment by alschmut** [comment]
So I just run the visualisation for this branch, and found these two bugs:

### undefined mode
- Select `Multiple` mode
- Open the mapSelection
- Unselect all maps or click button `None`
=> now no mode is selected, the loadingMapGif is showing and
=> the mapSelection disappears, when clicking outside of the mapSelection menu
Comment: While I kinda like that it's not setting the only available unary metric for all three metrics as it did before, we should find a solution here that behaves like before or better.

### preselected multiple mode
- Select `Multiple` mode
- Import two or more files
=> now sometimes the `multiple` mode is still selected, although it should by default select `single` mode after importing new files

File: visualization/app/codeCharta/model/files.ts (hunk 24 above)
Comment (alschmut) [reply visualization/app/codeCharta/model/files.ts]: I think it makes sense to postpone the renaming to another PR, as this PR is already pretty huge.

**Comment by JaiJoPa** [comment]
> After implementing the requested changes and fixing the selection of zero files in Multiple mode, I am fine with merging this PR 🎉

I like the fix! Thanks @alschmut
